oak/server/rust/oak_runtime/src/node/http/mod.rs

```diff
@@ -0,0 +1,438 @@
+#[derive(Clone)]
+struct HttpRequestHandler {
+    /// Reference to the Runtime in the context of this HTTP server pseudo-Node.
+    runtime: RuntimeProxy,
+    /// Channel handle used for writing HTTP invocations.
+    invocation_channel: oak_abi::Handle,
+}
+
+impl HttpRequestHandler {
+    fn handle(
+        &self,
+        request: Request<Body>,
+    ) -> Pin<Box<dyn Future<Output = Result<Response<Body>, OakStatus>> + Send + Sync>> {
+        let handler = self.clone();
+        let url = request.uri().to_string();
+        let method = request.method().as_str().to_string();
+
+        let future = async move {
+            let oak_label = get_oak_label(&request)?;
+            // TODO - cannot use features::block_on(hyper::body::to_bytes(request.into_body()))
+            let body = vec![]; 
+            info!(
+                "handling HTTP/2 request; request size: {} bytes, label: {:?}",
+                body.len(),
+                oak_label
+            );
+
+            let http_request = HttpRequest {
+                url,
+                method,
+                body,
+                headers: HashMap::new(), // TODO - get headers from the request
+            };
+
+            debug!("inject the request into the Oak Node");
+            let response = handler
+                .inject_http_request(http_request, &oak_label)
+                .map_err(|_| OakStatus::ErrInternal)?;
+
+            Ok(response.to_response())
+        };
+
+        Box::pin(future)
+    }
+
+    fn inject_http_request(
+        &self,
+        request: HttpRequest,
+        label: &Label,
+    ) -> Result<HttpResponseIterator, ()> {
+        // Create a pair of temporary channels to pass the HTTP request and to receive the response.
+        let pipe = Pipe::new(&self.runtime.clone(), label)?;
+
+        // Serialize HTTP/2 request into a message.
+        let mut message = crate::NodeMessage {
+            data: vec![],
+            handles: vec![],
+        };
+
+        request.encode(&mut message.data).map_err(|err| {
+            error!("Couldn't serialize HttpRequest message: {}", err);
+        })?;
+
+        // Put the HTTP request message inside the per-invocation request channel.
+        pipe.insert_message(&self.runtime, message)?;
+
+        // Send an invocation message (with attached handles) to the Oak Node.
+        pipe.send_invocation(&self.runtime, self.invocation_channel)?;
+
+        // Close all local handles except for the one that allows reading responses.
+        pipe.close(&self.runtime);
+
+        Ok(HttpResponseIterator {
+            runtime: self.runtime.clone(),
+            response_reader: pipe.response_reader,
+        })
+    }
+}
+
+// A pair of temporary channels to pass the HTTP request and to receive the response.
+struct Pipe {
+    request_writer: oak_abi::Handle,
+    request_reader: oak_abi::Handle,
+    response_writer: oak_abi::Handle,
+    response_reader: oak_abi::Handle,
+}
```

Comment:
Encapsulating this stuff is a good idea, but I wonder if it could be streamlined more, but still remain re-usable by both the HTTP server pseudo-Node and the gRPC server pseudo-Node.

In either case the pseudo-Node wants to:
 - create a (req, rsp) pair of channels
 - write a message to the req channel, then close the req-write handle
 - write an invocation message, holding req-read and rsp-write handles, to an invocation channel
 - close the local req-read and rsp-write handles
 - use the rsp-read handle to block on, waiting for response(s)

That kind of implies a standalone method with a signature that takes `(runtime: RuntimeProxy, req: NodeMessage, invocation_channel: oak_abi::Handle)` as input, and returns `rsp_reader: oak_abi::Handle` as output. 

Could the existing `GrpcInvocationHandler::inject_grpc_request` be refactored (in a precursor PR) to be a thin wrapper around such a function, and then re-used in this PR?

Reply:
A warning, though: refactoring the gRPC server p-N code is likely to interact with #1249, so you might need to sync with @conradgrobler & @ipetr0v